File: doc/example_3.f
C Example 3:
C In this example, we construct a ball-and-stick water molecule, using
C the camera and lights developped in Example 2. 

C Subroutine to ease checking of return codes
      subroutine errchk( ival )
      if (ival.ne.1) write(6,*) 'ERROR!'
      return
      end

C Subroutine to create the water molecule
      subroutine mkh2o()
      integer popen, pscale, psphr, pgbclr, pclose, ptrans, pchild,
     $     pplyln

C Atomic information- radii are 1/4 of Van der Waals
      real hrad, orad
      data hrad, orad/0.3, 0.35/
      real hclr(4), oclr(4)
      data hclr/1.0, 1.0, 1.0, 1.0/
      data oclr/1.0, 0.0, 0.0, 1.0/
      real h1loc(3), h2loc(3), oloc(3)
      data h1loc/0.854821, -.487790, 0.0/
      data h2loc/-0.854821, -.487790, 0.0/
      data oloc/0.0, 0.121948, 0.0/

C Polyline to join the atoms
      real lnclr(4)
      data lnclr/0.0, 0.0, 1.0, 1.0/
      real lncrd(3,3)
      data lncrd/
     $     0.854821, -.487790, 0.0,
     $     0.0, 0.121948, 0.0,
     $     -0.854821, -.487790, 0.0 /
      
C Make spheres of appropriate radius and color for elements.
C Note that the order in which attributes, transformations, and
C GOBs are inserted into a named GOB is irrelevant.
      call errchk( popen('hydrogen') )
      call errchk( pscale(hrad) )
      call errchk( psphr() )
      call errchk( pgbclr( 0, hclr(1), hclr(2), hclr(3), hclr(4) ) )
      call errchk( pclose() )

      call errchk( popen('oxygen') )
      call errchk( pscale(orad) )
      call errchk( psphr() )
      call errchk( pgbclr( 0, oclr(1), oclr(2), oclr(3), oclr(4) ) )
      call errchk( pclose() )

C Open the outer GOB, and insert in it three nameless subGOBs.  Each
C contains one atom and the transformation appropriate to put it in
C the proper place.  
      call errchk( popen('h2o') )

      call errchk( popen(' ') )
      call errchk( ptrans( h1loc(1), h1loc(2), h1loc(3) ) )
      call errchk( pchild('hydrogen') )
      call errchk( pclose() )

      call errchk( popen(' ') )
      call errchk( ptrans( h2loc(1), h2loc(2), h2loc(3) ) )
      call errchk( pchild('hydrogen') )
      call errchk( pclose() )

      call errchk( popen(' ') )
      call errchk( ptrans( oloc(1), oloc(2), oloc(3) ) )
      call errchk( pchild('oxygen') )
      call errchk( pclose() )

C Insert a polyline connecting the atoms.  The color would apply
C to the atoms as well, but it will be superceded by their own
C color attribute.
      call errchk( pgbclr(0, lnclr(1), lnclr(2), lnclr(3), lnclr(4) ) )
      call errchk( pplyln(0, 0, 3, lncrd, 0 ) )

C Close the outer GOB
      call errchk( pclose() )

      return
      end

      program ex3
      integer pintrn, popen, pclose, psnap, pshtdn, pamblt, plight,
     $     pcamra

C Light specification information
      real ltloc(3), ltclr(4), ambclr(4)
      data ltloc/0.0, 20.0, 20.0/
      data ltclr/0.7, 0.7, 0.7, 1.0/
      data ambclr/0.3, 0.3, 0.3, 1.0/

C Camera specification information
      real lookfm(3), lookat(3), up(3)
      data lookfm/0.0, 0.0, 10.0/
      data lookat/0.0, 0.0, 0.0/
      data up/0.0, 1.0, 0.0/
      real fovea, hither, yon
      data fovea, hither, yon/ 53.0, -5.0, -15.0 /

C Create and initialize the renderer
      call errchk( pintrn('myrenderer','p3d','example_3.p3d',
     $     'This string will be ignored') )

C Create the lights and camera
      call errchk( popen('mylights') )
      call errchk( pamblt( 0, ambclr ) )
      call errchk( plight( ltloc, 0, ltclr ) )
      call errchk( pclose() )
      call errchk( pcamra('mycamera', lookfm, lookat, up, fovea,
     $     hither, yon ) )

C Create the water molecule
      call mkh2o()

C Render the model
      call errchk( psnap('h2o','mylights','mycamera') )

C Shut down the renderer
      call errchk( pshtdn() )

      stop
      end

File: doc/example_3.c
/* Example 3:
 * In this example, we construct a ball-and-stick water molecule, using
 * the camera and lights developped in Example 2. 
 */

#include <stdio.h>
#include <p3dgen.h>
#include <drawp3d.h>

/* Macro to ease checking of return codes */
#define ERRCHK( string ) if (!(string)) fprintf(stderr,"ERROR!\n")

/* Light specification information */
static P_Point lightloc= { 0.0, 20.0, 20.0 };
static P_Color lightcolor= { P3D_RGB, 0.7, 0.7, 0.7, 1.0 };
static P_Color ambcolor= { P3D_RGB, 0.3, 0.3, 0.3, 1.0 };

/* Camera specification information */
static P_Point lookfrom= { 0.0, 0.0, 10.0 };
static P_Point lookat= { 0.0, 0.0, 0.0 };
static P_Vector up= { 0.0, 1.0, 0.0 };
static float fovea= 53.0;
static float hither= -5.0;
static float yon= -15.0;

/* Atomic information */
static float h_radius= 0.3; /* 1/4 of Van der Waals */
static P_Color h_color= { P3D_RGB, 1.0, 1.0, 1.0, 1.0 };
static float o_radius= 0.35; /* 1/4 of Van der Waals */
static P_Color o_color= { P3D_RGB, 1.0, 0.0, 0.0, 1.0 };
static P_Point h1_loc= { 0.854821, -.487790, 0.0 };
static P_Point h2_loc= { -0.854821, -.487790, 0.0 };
static P_Point o_loc= { 0.0, 0.121948, 0.0 };

/* Polyline to join the atoms */
static P_Color pline_color= { P3D_RGB, 0.0, 0.0, 1.0, 1.0 };
static float pline_coords[]= {
  0.854821, -.487790, 0.0,
  0.0, 0.121948, 0.0,
  -0.854821, -.487790, 0.0 
  };

static void create_h2o()
/* This routine creates the water molecule- a GOB named 'h2o' */
{
  /* Make spheres of appropriate radius and color for elements.
   * Note that the order in which attributes, transformations, and
   * GOBs are inserted into a named GOB is irrelevant.
   */
  ERRCHK( dp_open("hydrogen") );
  ERRCHK( dp_scale(h_radius) );
  ERRCHK( dp_sphere() );
  ERRCHK( dp_gobcolor(&h_color) );
  ERRCHK( dp_close() );

  ERRCHK( dp_open("oxygen") );
  ERRCHK( dp_scale(o_radius) );
  ERRCHK( dp_sphere() );
  ERRCHK( dp_gobcolor(&o_color) );
  ERRCHK( dp_close() );

  /* Open the outer GOB, and insert in it three nameless subGOBs.  Each
   * contains one atom and the transformation appropriate to put it in
   * the proper place.  
   */
  ERRCHK( dp_open("h2o") );

  ERRCHK( dp_open("") );
  ERRCHK( dp_translate( h1_loc.x, h1_loc.y, h1_loc.z ) );
  ERRCHK( dp_child("hydrogen") );
  ERRCHK( dp_close() );

  ERRCHK( dp_open("") );
  ERRCHK( dp_translate( h2_loc.x, h2_loc.y, h2_loc.z ) );
  ERRCHK( dp_child("hydrogen") );
  ERRCHK( dp_close() );

  ERRCHK( dp_open("") );
  ERRCHK( dp_translate( o_loc.x, o_loc.y, o_loc.z ) );
  ERRCHK( dp_child("oxygen") );
  ERRCHK( dp_close() );

  /* Insert a polyline connecting the atoms.  The color would apply
   * to the atoms as well, but it will be superceded by their own
   * color attribute.
   */
  ERRCHK( dp_gobcolor(&pline_color) );
  ERRCHK( dp_polyline( P3D_CVTX, P3D_RGB, pline_coords, 3 ) );

  /* Close the outer GOB */
  ERRCHK( dp_close() );
}

main()
{
  /* Initialize renderer */
  ERRCHK( dp_init_ren("myrenderer","p3d","example_3.p3d",
		      "This string will be ignored") );
  
  /* Create my lights and camera */
  ERRCHK( dp_open("mylights") );
  ERRCHK( dp_ambient(&ambcolor) );
  ERRCHK( dp_light(&lightloc, &lightcolor) );
  ERRCHK( dp_close() );
  ERRCHK( dp_camera( "mycamera", &lookfrom, &lookat, &up, fovea,
		    hither, yon ) );

  /* Create the water molecule */
  create_h2o();

  /* Render the model */
  ERRCHK( dp_snap("h2o","mylights","mycamera") );

  /* Shut down renderer */
  ERRCHK( dp_shutdown() );
}
